sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -597,6 +597,16 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     }
   }
 
+  override def alterTableSchema(db: String, table: String, schema: StructType): Unit = withClient {
+    requireTableExists(db, table)
+    val rawTable = getRawTable(db, table)
+    val withNewSchema = rawTable.copy(schema = schema)
+    // Add table metadata such as table schema, partition columns, etc. to table properties.
+    val updatedTable = withNewSchema.copy(
+      properties = withNewSchema.properties ++ tableMetaToTableProps(withNewSchema))
+    client.alterTable(updatedTable)
```

Comment:
one more thing, if you look at `def createDataSourceTable`, we need to care about one special case: when saving table to hive, we should try catch it, if it fails, convert it to non-hive-compatible format(set schema to Nil) and save again. We should follow it here, in case the schema we wanna alter to is not hive compatible.